### PR TITLE
[photonics] [C++] Support 'get_state' API

### PR DESCRIPTION
**Pull request by khalatepradnya**
Towards: 'photonics' target support

* Added `get_state` API in C++ frontend
* Python bindings to be added after PR #2055

### CODE REVIEW

**Comment by khalatepradnya** [comment]
@Omar-ORCA for visibility

**Comment by thorinf-orca** [comment]
Hi Pradnya, 

This looks great, thanks for working on this for us. Do you have any guidance on how we could do our Python bindings?

Thanks,

Thorin

File: runtime/cudaq/qis/managers/photonics/PhotonicsExecutionManager.cpp [hunk 1]
```diff
@@ -19,6 +19,68 @@
...
+  std::complex<double>
+  getAmplitude(const std::vector<int> &basisState) override {
+    /// TODO: Check basisState.size() matches qudit count
+
+    // Convert the basis state to an index value
+    const std::size_t idx = std::accumulate(
+        std::make_reverse_iterator(basisState.end()),
+        std::make_reverse_iterator(basisState.begin()), 0ull,
+        [](std::size_t acc, int bit) { return (acc << 1) + bit; });
+    return state[idx];
```
Comment (1tnguyen) [inline runtime/cudaq/qis/managers/photonics/PhotonicsExecutionManager.cpp:45]: I thought this logic was for binary states (`basisState` is a vector of 0 or 1). For qudit states of dimension `d`, it should be `acc * d + bit`.

Comment (khalatepradnya) [reply runtime/cudaq/qis/managers/photonics/PhotonicsExecutionManager.cpp]: Thanks, Thien! Addressed this in f6113fd40becb08d47c273518794ae1524da640c.

**Comment by khalatepradnya** [comment]
> Hi Pradnya,
> 
> This looks great, thanks for working on this for us. Do you have any guidance on how we could do our Python bindings?
> 
> Thanks,
> 
> Thorin

Thanks, Thorin! You will need the changes in PR# https://github.com/NVIDIA/cuda-quantum/pull/2055 to add Python bindings for `get_state()` API for the `photonics` target.

**Review by 1tnguyen — APPROVED** [review]
LGTM

**Pull request merged**